File: 01-reference-agent/cpp/yahoo.cpp
#include "yahoo.hpp"

#include <nlohmann/json.hpp>

#include <algorithm>
#include <chrono>
#include <cctype>
#include <curl/curl.h>
#include <filesystem>
#include <fstream>
#include <sstream>
#include <stdexcept>
#include <thread>

using json = nlohmann::json;
namespace fs = std::filesystem;

namespace {

constexpr const char* kUserAgent =
    "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) "
    "AppleWebKit/537.36 (KHTML, like Gecko) Chrome/124.0.0.0 Safari/537.36";

const char* kYahooHosts[] = {
    "https://query1.finance.yahoo.com/v1/finance/search",
    "https://query2.finance.yahoo.com/v1/finance/search",
};

size_t write_cb(char* ptr, size_t size, size_t nmemb, void* userdata) {
    auto* out = static_cast<std::string*>(userdata);
    out->append(ptr, size * nmemb);
    return size * nmemb;
}

std::string now_iso() {
    using namespace std::chrono;
    const auto secs =
        duration_cast<seconds>(system_clock::now().time_since_epoch()).count();
    return std::to_string(secs);
}

fs::path cache_path() {
    return fs::path(example_root()) / "stories" / "stories_cache.json";
}

fs::path stories_dir() {
    return fs::path(example_root()) / "stories";
}

json load_cache() {
    std::ifstream in(cache_path());
    if (!in) {
        return json{{"updated_at", nullptr}, {"tickers", json::object()}, {"last_pair", nullptr}};
    }
    try {
        json data;
        in >> data;
        if (!data.is_object()) {
            throw std::runtime_error("bad cache");
        }
        if (!data.contains("tickers") || !data["tickers"].is_object()) {
            data["tickers"] = json::object();
        }
        return data;
    } catch (...) {
        return json{{"updated_at", nullptr}, {"tickers", json::object()}, {"last_pair", nullptr}};
    }
}

void save_cache(json cache) {
    cache["updated_at"] = now_iso();
    fs::create_directories(stories_dir());
    std::ofstream out(cache_path());
    out << cache.dump(2) << '\n';
}

std::optional<TickerBlock> get_cached_ticker(const std::string& ticker) {
    const std::string symbol = normalize_ticker(ticker);
    if (symbol.empty()) {
        return std::nullopt;
    }
    auto cache = load_cache();
    if (!cache["tickers"].contains(symbol)) {
        return std::nullopt;
    }
    const auto& entry = cache["tickers"][symbol];
    if (!entry.contains("stories") || !entry["stories"].is_array() ||
        entry["stories"].empty()) {
        return std::nullopt;
    }
    TickerBlock block;
    block.ticker = symbol;
    block.name = entry.value("name", symbol);
    if (block.name.empty()) {
        block.name = symbol;
    }
    block.from_cache = true;
    int n = 0;
    for (const auto& s : entry["stories"]) {
        if (n >= 2) {
            break;
        }
        Story story;
        story.title = s.value("title", "");
        story.publisher = s.value("publisher", "");
        story.link = s.value("link", "");
        story.uuid = s.value("uuid", "");
        if (!story.title.empty()) {
            block.stories.push_back(std::move(story));
            ++n;
        }
    }
    if (block.stories.empty()) {
        return std::nullopt;
    }
    return block;
}

void remember_ticker(const std::string& symbol, const std::string& name,
                     const std::vector<Story>& stories) {
    if (stories.empty()) {
        return;
    }
    auto cache = load_cache();
    json arr = json::array();
    for (size_t i = 0; i < stories.size() && i < 2; ++i) {
        arr.push_back({{"title", stories[i].title},
                       {"publisher", stories[i].publisher},
                       {"link", stories[i].link},
                       {"uuid", stories[i].uuid}});
    }
    cache["tickers"][symbol] = {{"name", name.empty() ? symbol : name},
                                {"stories", arr},
                                {"cached_at", now_iso()}};
    save_cache(std::move(cache));
}

void remember_pair(const std::string& t1, const std::string& t2,
                   const std::vector<TickerBlock>& results) {
    if (results.size() != 2) {
        return;
    }
    for (const auto& r : results) {
        if (r.stories.empty()) {
            return;
        }
    }
    auto cache = load_cache();
    cache["last_pair"] = {{"ticker1", normalize_ticker(t1)},
                          {"ticker2", normalize_ticker(t2)}};
    for (const auto& block : results) {
        if (block.stories.empty() || block.from_cache) {
            continue;
        }
        const std::string symbol = normalize_ticker(block.ticker);
        if (symbol.empty()) {
            continue;
        }
        json arr = json::array();
        for (size_t i = 0; i < block.stories.size() && i < 2; ++i) {
            arr.push_back({{"title", block.stories[i].title},
                           {"publisher", block.stories[i].publisher},
                           {"link", block.stories[i].link},
                           {"uuid", block.stories[i].uuid}});
        }
        cache["tickers"][symbol] = {
            {"name", block.name.empty() ? symbol : block.name},
            {"stories", arr},
            {"cached_at", now_iso()}};
    }
    save_cache(std::move(cache));
}

std::string http_get(const std::string& url, long* status_out) {
    std::string body;
    CURL* curl = curl_easy_init();
    if (!curl) {
        throw std::runtime_error("curl_easy_init failed");
    }
    curl_easy_setopt(curl, CURLOPT_URL, url.c_str());
    curl_easy_setopt(curl, CURLOPT_USERAGENT, kUserAgent);
    curl_easy_setopt(curl, CURLOPT_ACCEPT_ENCODING, "");
    curl_easy_setopt(curl, CURLOPT_FOLLOWLOCATION, 1L);
    curl_easy_setopt(curl, CURLOPT_TIMEOUT, 20L);
    curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, write_cb);
    curl_easy_setopt(curl, CURLOPT_WRITEDATA, &body);
    struct curl_slist* headers = nullptr;
    headers = curl_slist_append(headers, "Accept: application/json");
    headers = curl_slist_append(headers, "Accept-Language: en-US,en;q=0.9");
    curl_easy_setopt(curl, CURLOPT_HTTPHEADER, headers);
    const CURLcode rc = curl_easy_perform(curl);
    long status = 0;
    curl_easy_getinfo(curl, CURLINFO_RESPONSE_CODE, &status);
    curl_slist_free_all(headers);
    curl_easy_cleanup(curl);
    if (status_out) {
        *status_out = status;
    }
    if (rc != CURLE_OK) {
        throw std::runtime_error(curl_easy_strerror(rc));
    }
    return body;
}

json yahoo_get_json(const std::string& url) {
    std::string last_err = "Yahoo request failed";
    for (int attempt = 0; attempt < 3; ++attempt) {
        try {
            long status = 0;
            const std::string body = http_get(url, &status);
            if (status == 429 || status == 503) {
                last_err = "HTTP " + std::to_string(status);
                if (attempt < 2) {
                    std::this_thread::sleep_for(
                        std::chrono::milliseconds(1500 * (attempt + 1)));
                    continue;
                }
                throw std::runtime_error(last_err);
            }
            if (status < 200 || status >= 300) {
                throw std::runtime_error("HTTP " + std::to_string(status));
            }
            return json::parse(body);
        } catch (const std::exception& ex) {
            last_err = ex.what();
            if (attempt < 2) {
                std::this_thread::sleep_for(std::chrono::seconds(1));
                continue;
            }
        }
    }
    throw std::runtime_error(last_err);
}

std::optional<TickerBlock> parse_search_payload(const std::string& symbol,
                                                const json& payload, int count) {
    std::string name;
    if (payload.contains("quotes") && payload["quotes"].is_array() &&
        !payload["quotes"].empty()) {
        const auto& q0 = payload["quotes"][0];
        name = q0.value("shortname", "");
        if (name.empty()) {
            name = q0.value("longname", "");
        }
    }
    TickerBlock block;
    block.ticker = symbol;
    block.from_cache = false;
    if (payload.contains("news") && payload["news"].is_array()) {
        int n = 0;
        for (const auto& item : payload["news"]) {
            if (n >= count) {
                break;
            }
            Story s;
            s.title = item.value("title", "");
            // trim
            while (!s.title.empty() && std::isspace(static_cast<unsigned char>(s.title.front()))) {
                s.title.erase(s.title.begin());
            }
            while (!s.title.empty() && std::isspace(static_cast<unsigned char>(s.title.back()))) {
                s.title.pop_back();
            }
            if (s.title.empty()) {
                continue;
            }
            s.publisher = item.value("publisher", "");
            s.link = item.value("link", "");
            s.uuid = item.value("uuid", "");
            block.stories.push_back(std::move(s));
            ++n;
        }
    }
    if (block.stories.empty()) {
        return std::nullopt;
    }
    block.name = name.empty() ? symbol : name;
    return block;
}

TickerBlock fetch_stories_for_ticker(const std::string& ticker, int count) {
    const std::string symbol = normalize_ticker(ticker);
    if (symbol.empty()) {
        return {"", "", {}, "Ticker is empty.", false};
    }
    count = std::max(1, count);
    const std::string variants[] = {
        "q=" + symbol + "&quotesCount=1&newsCount=" + std::to_string(count) +
            "&enableFuzzyQuery=false&newsQueryId=news_cie_vespa&lang=en-US&region=US",
        "q=" + symbol + "&quotesCount=1&newsCount=" + std::to_string(count) +
            "&lang=en-US&region=US",
    };
    std::string last_error = "No recent stories found for " + symbol + ".";
    for (const char* host : kYahooHosts) {
        for (const auto& params : variants) {
            const std::string url = std::string(host) + "?" + params;
            try {
                const json payload = yahoo_get_json(url);
                auto parsed = parse_search_payload(symbol, payload, count);
                if (!parsed) {
                    last_error = "No recent stories found for " + symbol + ".";
                    continue;
                }
                remember_ticker(symbol, parsed->name, parsed->stories);
                return *parsed;
            } catch (const std::exception& ex) {
                const std::string msg = ex.what();
                if (msg.rfind("HTTP ", 0) == 0) {
                    last_error = "Yahoo Finance " + msg + " for " + symbol + ".";
                } else {
                    last_error = "Yahoo Finance request failed for " + symbol + ": " + msg;
                }
            }
        }
    }
    if (auto cached = get_cached_ticker(symbol)) {
        cached->error = last_error + " Showing last saved headlines.";
        return *cached;
    }
    return {symbol, symbol, {}, last_error, false};
}

}  // namespace

std::string example_root() {
    return (fs::current_path() / "..").lexically_normal().string();
}

std::string normalize_ticker(const std::string& raw) {
    std::string out;
    for (unsigned char ch : raw) {
        if (std::isspace(ch)) {
            continue;
        }
        ch = static_cast<unsigned char>(std::toupper(ch));
        if ((ch >= 'A' && ch <= 'Z') || (ch >= '0' && ch <= '9') || ch == '.' ||
            ch == '-') {
            out.push_back(static_cast<char>(ch));
        }
    }
    return out;
}

std::optional<std::tuple<std::string, std::string, std::vector<TickerBlock>>>
get_last_pair_cached() {
    auto cache = load_cache();
    if (!cache.contains("last_pair") || cache["last_pair"].is_null() ||
        !cache["last_pair"].is_object()) {
        return std::nullopt;
    }
    const std::string t1 = normalize_ticker(cache["last_pair"].value("ticker1", ""));
    const std::string t2 = normalize_ticker(cache["last_pair"].value("ticker2", ""));
    if (t1.empty() || t2.empty()) {
        return std::nullopt;
    }
    std::vector<TickerBlock> blocks;
    for (const auto& symbol : {t1, t2}) {
        auto cached = get_cached_ticker(symbol);
        if (!cached) {
            return std::nullopt;
        }
        blocks.push_back(*cached);
    }
    return std::make_tuple(t1, t2, blocks);
}

FetchPairResult fetch_stories_for_tickers(const std::string& ticker1,
                                          const std::string& ticker2, int count) {
    std::string t1 = normalize_ticker(ticker1);
    if (t1.empty()) {
        t1 = kDefaultTicker1;
    }
    std::string t2 = normalize_ticker(ticker2);
    if (t2.empty()) {
        t2 = kDefaultTicker2;
    }
    auto first = fetch_stories_for_ticker(t1, count);
    std::this_thread::sleep_for(std::chrono::milliseconds(500));
    auto second = fetch_stories_for_ticker(t2, count);
    std::vector<TickerBlock> results{std::move(first), std::move(second)};
    remember_pair(t1, t2, results);
    FetchPairResult out;
    out.ticker1 = t1;
    out.ticker2 = t2;
    out.tickers = std::move(results);
    for (const auto& r : out.tickers) {
        if (!r.error.empty()) {
            out.errors.push_back(r.error);
        }
    }
    return out;
}

std::string format_stories_for_prompt(const std::vector<TickerBlock>& ticker_results) {
    std::ostringstream b;
    b << "Using only the recent Yahoo Finance headlines below, write a short "
         "market briefing that compares the two tickers. Cite story titles "
         "where helpful. Do not invent facts beyond what the headlines imply.\n\n";
    for (const auto& block : ticker_results) {
        const std::string ticker = block.ticker.empty() ? "?" : block.ticker;
        const std::string name = block.name.empty() ? ticker : block.name;
        b << "## " << ticker << " (" << name << ")\n";
        if (block.stories.empty()) {
            b << "- (no stories available)\n";
            if (!block.error.empty()) {
                b << "- note: " << block.error << "\n";
            }
        } else {
            for (size_t i = 0; i < block.stories.size(); ++i) {
                const auto& s = block.stories[i];
                const std::string title = s.title.empty() ? "(untitled)" : s.title;
                const std::string publisher =
                    s.publisher.empty() ? "unknown" : s.publisher;
                b << (i + 1) << ". " << title << " — " << publisher << "\n";
            }
        }
        b << "\n";
    }
    std::string text = b.str();
    while (!text.empty() && (text.back() == '\n' || text.back() == ' ')) {
        text.pop_back();
    }
    return text;
}

File: 01-reference-agent/go/yahoo.go
package main

import (
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"os"
	"path/filepath"
	"regexp"
	"strings"
	"time"
)

const (
	defaultTicker1 = "NVDA"
	defaultTicker2 = "SPCX"
	userAgent     = "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) " +
		"AppleWebKit/537.36 (KHTML, like Gecko) Chrome/124.0.0.0 Safari/537.36"
)

var yahooSearchHosts = []string{
	"https://query1.finance.yahoo.com/v1/finance/search",
	"https://query2.finance.yahoo.com/v1/finance/search",
}

var tickerCleaner = regexp.MustCompile(`[^A-Z0-9.\-]`)

type story struct {
	Title     string `json:"title"`
	Publisher string `json:"publisher"`
	Published string `json:"published"`
	Link      string `json:"link"`
	UUID      string `json:"uuid"`
}

type tickerBlock struct {
	Ticker    string  `json:"ticker"`
	Name      string  `json:"name"`
	Stories   []story `json:"stories"`
	Source    string  `json:"source,omitempty"`
	Error     string  `json:"error,omitempty"`
	FromCache bool    `json:"from_cache"`
	CachedAt  string  `json:"cached_at,omitempty"`
}

type cacheTickerEntry struct {
	Name     string  `json:"name"`
	Stories  []story `json:"stories"`
	CachedAt string  `json:"cached_at"`
}

type storiesCache struct {
	UpdatedAt *string                     `json:"updated_at"`
	Tickers   map[string]cacheTickerEntry `json:"tickers"`
	LastPair  *struct {
		Ticker1 string `json:"ticker1"`
		Ticker2 string `json:"ticker2"`
	} `json:"last_pair"`
}

type fetchPairResult struct {
	Tickers []tickerBlock
	OK      bool
	Errors  []string
	Ticker1 string
	Ticker2 string
}

func exampleRoot() string {
	return filepath.Clean(filepath.Join(".."))
}

func cachePath() string {
	return filepath.Join(exampleRoot(), "stories", "stories_cache.json")
}

func storiesDir() string {
	return filepath.Join(exampleRoot(), "stories")
}

func normalizeTicker(raw string) string {
	s := strings.ToUpper(strings.TrimSpace(raw))
	return tickerCleaner.ReplaceAllString(s, "")
}

func nowISO() string {
	return time.Now().UTC().Format("2006-01-02T15:04:05Z")
}

func loadCache() storiesCache {
	empty := storiesCache{Tickers: map[string]cacheTickerEntry{}}
	data, err := os.ReadFile(cachePath())
	if err != nil {
		return empty
	}
	var c storiesCache
	if err := json.Unmarshal(data, &c); err != nil {
		return empty
	}
	if c.Tickers == nil {
		c.Tickers = map[string]cacheTickerEntry{}
	}
	return c
}

func saveCache(c storiesCache) error {
	ts := nowISO()
	c.UpdatedAt = &ts
	if err := os.MkdirAll(storiesDir(), 0o755); err != nil {
		return err
	}
	raw, err := json.MarshalIndent(c, "", "  ")
	if err != nil {
		return err
	}
	return os.WriteFile(cachePath(), append(raw, '\n'), 0o644)
}

func getCachedTicker(ticker string) *tickerBlock {
	symbol := normalizeTicker(ticker)
	if symbol == "" {
		return nil
	}
	entry, ok := loadCache().Tickers[symbol]
	if !ok || len(entry.Stories) == 0 {
		return nil
	}
	stories := entry.Stories
	if len(stories) > 2 {
		stories = stories[:2]
	}
	name := entry.Name
	if name == "" {
		name = symbol
	}
	return &tickerBlock{
		Ticker:    symbol,
		Name:      name,
		Stories:   append([]story(nil), stories...),
		Source:    "cache",
		FromCache: true,
		CachedAt:  entry.CachedAt,
	}
}

func getLastPairCached() (t1, t2 string, blocks []tickerBlock, ok bool) {
	c := loadCache()
	if c.LastPair == nil {
		return "", "", nil, false
	}
	t1 = normalizeTicker(c.LastPair.Ticker1)
	t2 = normalizeTicker(c.LastPair.Ticker2)
	if t1 == "" || t2 == "" {
		return "", "", nil, false
	}
	blocks = make([]tickerBlock, 0, 2)
	for _, symbol := range []string{t1, t2} {
		cached := getCachedTicker(symbol)
		if cached == nil {
			return "", "", nil, false
		}
		blocks = append(blocks, *cached)
	}
	return t1, t2, blocks, true
}

func rememberTicker(symbol, name string, stories []story) {
	if len(stories) == 0 {
		return
	}
	if len(stories) > 2 {
		stories = stories[:2]
	}
	c := loadCache()
	if name == "" {
		name = symbol
	}
	c.Tickers[symbol] = cacheTickerEntry{
		Name:     name,
		Stories:  append([]story(nil), stories...),
		CachedAt: nowISO(),
	}
	_ = saveCache(c)
}

func rememberPair(ticker1, ticker2 string, results []tickerBlock) {
	if len(results) != 2 {
		return
	}
	for _, r := range results {
		if len(r.Stories) == 0 {
			return
		}
	}
	c := loadCache()
	c.LastPair = &struct {
		Ticker1 string `json:"ticker1"`
		Ticker2 string `json:"ticker2"`
	}{Ticker1: normalizeTicker(ticker1), Ticker2: normalizeTicker(ticker2)}
	for _, block := range results {
		if len(block.Stories) == 0 || block.FromCache {
			continue
		}
		symbol := normalizeTicker(block.Ticker)
		if symbol == "" {
			continue
		}
		stories := block.Stories
		if len(stories) > 2 {
			stories = stories[:2]
		}
		name := block.Name
		if name == "" {
			name = symbol
		}
		c.Tickers[symbol] = cacheTickerEntry{
			Name:     name,
			Stories:  append([]story(nil), stories...),
			CachedAt: nowISO(),
		}
	}
	_ = saveCache(c)
}

func envKey(name string) string {
	return strings.TrimSpace(os.Getenv(name))
}

func commonStory(title, publisher, published, link, uuid string) *story {
	title = strings.TrimSpace(title)
	if title == "" {
		return nil
	}
	return &story{
		Title:     title,
		Publisher: strings.TrimSpace(publisher),
		Published: strings.TrimSpace(published),
		Link:      strings.TrimSpace(link),
		UUID:      strings.TrimSpace(uuid),
	}
}

func tickerOk(symbol, name string, stories []story, source string) tickerBlock {
	if name == "" {
		name = symbol
	}
	return tickerBlock{Ticker: symbol, Name: name, Stories: stories, Source: source}
}

func getJSON(rawURL string, extraHeaders map[string]string, sleepBefore time.Duration) (any, error) {
	if sleepBefore > 0 {
		time.Sleep(sleepBefore)
	}
	client := &http.Client{Timeout: 20 * time.Second}
	var lastErr error
	for attempt := 0; attempt < 3; attempt++ {
		req, err := http.NewRequest(http.MethodGet, rawURL, nil)
		if err != nil {
			return nil, err
		}
		req.Header.Set("User-Agent", userAgent)
		req.Header.Set("Accept", "application/json")
		req.Header.Set("Accept-Language", "en-US,en;q=0.9")
		for k, v := range extraHeaders {
			req.Header.Set(k, v)
		}
		res, err := client.Do(req)
		if err != nil {
			lastErr = err
			time.Sleep(time.Second)
			continue
		}
		body, readErr := io.ReadAll(res.Body)
		res.Body.Close()
		if readErr != nil {
			lastErr = readErr
			continue
		}
		if res.StatusCode == 429 {
			return nil, fmt.Errorf("HTTP %d", res.StatusCode)
		}
		if res.StatusCode == 503 {
			lastErr = fmt.Errorf("HTTP %d", res.StatusCode)
			if attempt < 2 {
				time.Sleep(time.Duration(1500*(attempt+1)) * time.Millisecond)
				continue
			}
			return nil, lastErr
		}
		if res.StatusCode < 200 || res.StatusCode >= 300 {
			return nil, fmt.Errorf("HTTP %d", res.StatusCode)
		}
		var payload any
		if err := json.Unmarshal(body, &payload); err != nil {
			return nil, err
		}
		return payload, nil
	}
	if lastErr == nil {
		lastErr = fmt.Errorf("news request failed")
	}
	return nil, lastErr
}

func httpFail(label, symbol string, err error) string {
	msg := err.Error()
	if strings.HasPrefix(msg, "HTTP ") {
		return fmt.Sprintf("%s %s for %s.", label, msg, symbol)
	}
	return fmt.Sprintf("%s request failed for %s: %v", label, symbol, err)
}

func unixToISO(v any) string {
	switch t := v.(type) {
	case float64:
		if t <= 0 {
			return ""
		}
		return time.Unix(int64(t), 0).UTC().Format("2006-01-02T15:04:05Z")
	case json.Number:
		n, err := t.Int64()
		if err != nil || n <= 0 {
			return ""
		}
		return time.Unix(n, 0).UTC().Format("2006-01-02T15:04:05Z")
	default:
		s := strings.TrimSpace(asString(v))
		if s == "" {
			return ""
		}
		if ts, err := time.Parse(time.RFC3339, s); err == nil {
			return ts.UTC().Format("2006-01-02T15:04:05Z")
		}
		if ts, err := time.Parse("2006-01-02T15:04:05Z", s); err == nil {
			return ts.UTC().Format("2006-01-02T15:04:05Z")
		}
		return s
	}
}

func asString(v any) string {
	if v == nil {
		return ""
	}
	switch t := v.(type) {
	case string:
		return t
	default:
		return fmt.Sprint(t)
	}
}

func parseSearchPayload(symbol string, payload map[string]any, count int) *tickerBlock {
	name := ""
	if quotes, ok := payload["quotes"].([]any); ok && len(quotes) > 0 {
		if q0, ok := quotes[0].(map[string]any); ok {
			name = strings.TrimSpace(asString(q0["shortname"]))
			if name == "" {
				name = strings.TrimSpace(asString(q0["longname"]))
			}
		}
	}
	stories := make([]story, 0, count)
	news, _ := payload["news"].([]any)
	for i, item := range news {
		if i >= count {
			break
		}
		m, ok := item.(map[string]any)
		if !ok {
			continue
		}
		s := commonStory(asString(m["title"]), asString(m["publisher"]), unixToISO(m["providerPublishTime"]), asString(m["link"]), asString(m["uuid"]))
		if s != nil {
			stories = append(stories, *s)
		}
	}
	if len(stories) == 0 {
		return nil
	}
	block := tickerOk(symbol, name, stories, "yahoo")
	return &block
}

func fetchFinnhub(symbol string, count int) (tickerBlock, string, bool) {
	token := envKey("FINNHUB_API_KEY")
	if token == "" {
		return tickerBlock{}, "", false
	}
	end := time.Now().UTC()
	start := end.AddDate(0, 0, -7)
	q := url.Values{
		"symbol": {symbol},
		"from":   {start.Format("2006-01-02")},
		"to":     {end.Format("2006-01-02")},
		"token":  {token},
	}
	payload, err := getJSON("https://finnhub.io/api/v1/company-news?"+q.Encode(), nil, 0)
	if err != nil {
		return tickerBlock{}, httpFail("Finnhub", symbol, err), false
	}
	items, ok := payload.([]any)
	if !ok {
		return tickerBlock{}, fmt.Sprintf("Finnhub returned no stories for %s.", symbol), false
	}
	stories := make([]story, 0, count)
	for _, item := range items {
		if len(stories) >= count {
			break
		}
		m, ok := item.(map[string]any)
		if !ok {
			continue
		}
		s := commonStory(asString(m["headline"]), asString(m["source"]), unixToISO(m["datetime"]), asString(m["url"]), asString(m["id"]))
		if s != nil {
			stories = append(stories, *s)
		}
	}
	if len(stories) == 0 {
		return tickerBlock{}, fmt.Sprintf("No recent stories found for %s.", symbol), false
	}
	return tickerOk(symbol, symbol, stories, "finnhub"), "", true
}

func fetchMassive(symbol string, count int) (tickerBlock, string, bool) {
	token := envKey("MASSIVE_API_KEY")
	if token == "" {
		return tickerBlock{}, "", false
	}
	q := url.Values{
		"ticker": {symbol},
		"limit":  {fmt.Sprintf("%d", count)},
		"sort":   {"published_utc"},
		"order":  {"desc"},
	}
	payload, err := getJSON("https://api.massive.com/v2/reference/news?"+q.Encode(), map[string]string{
		"Authorization": "Bearer " + token,
	}, 0)
	if err != nil {
		return tickerBlock{}, httpFail("Massive", symbol, err), false
	}
	root, ok := payload.(map[string]any)
	if !ok {
		return tickerBlock{}, fmt.Sprintf("Massive returned no stories for %s.", symbol), false
	}
	results, _ := root["results"].([]any)
	stories := make([]story, 0, count)
	for _, item := range results {
		if len(stories) >= count {
			break
		}
		m, ok := item.(map[string]any)
		if !ok {
			continue
		}
		publisher := ""
		if pub, ok := m["publisher"].(map[string]any); ok {
			publisher = asString(pub["name"])
		} else {
			publisher = asString(m["publisher"])
		}
		link := asString(m["article_url"])
		if link == "" {
			link = asString(m["url"])
		}
		s := commonStory(asString(m["title"]), publisher, unixToISO(m["published_utc"]), link, asString(m["id"]))
		if s != nil {
			stories = append(stories, *s)
		}
	}
	if len(stories) == 0 {
		return tickerBlock{}, fmt.Sprintf("No recent stories found for %s.", symbol), false
	}
	return tickerOk(symbol, symbol, stories, "massive"), "", true
}

func fetchYahoo(symbol string, count int) (tickerBlock, string, bool) {
	if count < 1 {
		count = 1
	}
	variants := []url.Values{
		{
			"q":                {symbol},
			"quotesCount":      {"1"},
			"newsCount":        {fmt.Sprintf("%d", count)},
			"enableFuzzyQuery": {"false"},
			"newsQueryId":      {"news_cie_vespa"},
			"lang":             {"en-US"},
			"region":           {"US"},
		},
		{
			"q":           {symbol},
			"quotesCount": {"1"},
			"newsCount":   {fmt.Sprintf("%d", count)},
			"lang":        {"en-US"},
			"region":      {"US"},
		},
	}
	lastError := fmt.Sprintf("No recent stories found for %s.", symbol)
	for _, host := range yahooSearchHosts {
		for _, params := range variants {
			rawURL := host + "?" + params.Encode()
			payload, err := getJSON(rawURL, nil, time.Second)
			if err != nil {
				msg := err.Error()
				if strings.HasPrefix(msg, "HTTP ") {
					lastError = fmt.Sprintf("Yahoo Finance %s for %s.", msg, symbol)
					if msg == "HTTP 429" {
						return tickerBlock{}, lastError, false
					}
				} else {
					lastError = fmt.Sprintf("Yahoo Finance request failed for %s: %v", symbol, err)
				}
				continue
			}
			root, ok := payload.(map[string]any)
			if !ok {
				lastError = fmt.Sprintf("No recent stories found for %s.", symbol)
				continue
			}
			parsed := parseSearchPayload(symbol, root, count)
			if parsed == nil {
				lastError = fmt.Sprintf("No recent stories found for %s.", symbol)
				continue
			}
			return *parsed, "", true
		}
	}
	return tickerBlock{}, lastError, false
}

func fetchStoriesForTicker(ticker string, count int) tickerBlock {
	symbol := normalizeTicker(ticker)
	if symbol == "" {
		return tickerBlock{Error: "Ticker is empty."}
	}
	if count < 1 {
		count = 1
	}
	lastError := fmt.Sprintf("No recent stories found for %s.", symbol)
	if envKey("FINNHUB_API_KEY") != "" {
		if block, err, ok := fetchFinnhub(symbol, count); ok {
			rememberTicker(symbol, block.Name, block.Stories)
			return block
		} else if err != "" {
			lastError = err
		}
	}
	if envKey("MASSIVE_API_KEY") != "" {
		if block, err, ok := fetchMassive(symbol, count); ok {
			rememberTicker(symbol, block.Name, block.Stories)
			return block
		} else if err != "" {
			lastError = err
		}
	}
	if block, err, ok := fetchYahoo(symbol, count); ok {
		rememberTicker(symbol, block.Name, block.Stories)
		return block
	} else if err != "" {
		lastError = err
	}
	if cached := getCachedTicker(symbol); cached != nil {
		cached.Error = lastError + " Showing last saved headlines."
		return *cached
	}
	return tickerBlock{Ticker: symbol, Name: symbol, Error: lastError}
}

func fetchStoriesForTickers(ticker1, ticker2 string, count int) fetchPairResult {
	t1 := normalizeTicker(ticker1)
	if t1 == "" {
		t1 = defaultTicker1
	}
	t2 := normalizeTicker(ticker2)
	if t2 == "" {
		t2 = defaultTicker2
	}
	first := fetchStoriesForTicker(t1, count)
	if first.Source == "yahoo" || len(first.Stories) == 0 {
		time.Sleep(500 * time.Millisecond)
	}
	second := fetchStoriesForTicker(t2, count)
	results := []tickerBlock{first, second}
	rememberPair(t1, t2, results)
	errors := make([]string, 0, 2)
	for _, r := range results {
		if r.Error != "" {
			errors = append(errors, r.Error)
		}
	}
	return fetchPairResult{
		Tickers: results,
		OK:      len(errors) == 0,
		Errors:  errors,
		Ticker1: t1,
		Ticker2: t2,
	}
}

func formatStoriesForPrompt(tickerResults []tickerBlock) string {
	var b strings.Builder
	b.WriteString("Using only the recent headlines below, write a short ")
	b.WriteString("market briefing that compares the two tickers. Cite story titles ")
	b.WriteString("where helpful. Do not invent facts beyond what the headlines imply.\n\n")
	for _, block := range tickerResults {
		ticker := block.Ticker
		if ticker == "" {
			ticker = "?"
		}
		name := block.Name
		if name == "" {
			name = ticker
		}
		b.WriteString(fmt.Sprintf("## %s (%s)\n", ticker, name))
		if len(block.Stories) == 0 {
			b.WriteString("- (no stories available)\n")
			if block.Error != "" {
				b.WriteString(fmt.Sprintf("- note: %s\n", block.Error))
			}
		} else {
			for i, s := range block.Stories {
				title := s.Title
				if title == "" {
					title = "(untitled)"
				}
				publisher := s.Publisher
				if publisher == "" {
					publisher = "unknown"
				}
				b.WriteString(fmt.Sprintf("%d. %s — %s\n", i+1, title, publisher))
			}
		}
		b.WriteString("\n")
	}
	return strings.TrimSpace(b.String())
}
